tests/unit/select_connection_ioloop_tests.py

```diff
@@ -252,39 +293,140 @@ def on_timeout(self):
     SELECT_POLLER = 'poll'
 
 class IOLoopSimpleMessageTestCasetEPoll(IOLoopSimpleMessageTestCaseSelect):
+    ''' Same as IOLoopSimpleMessageTestCaseSelect but uses 'epoll' syscall '''
     SELECT_POLLER = 'epoll'
 
 class IOLoopSimpleMessageTestCasetKqueue(IOLoopSimpleMessageTestCaseSelect):
+    ''' Same as IOLoopSimpleMessageTestCaseSelect but uses 'kqueue' syscall '''
+    SELECT_POLLER = 'kqueue'
+
+
+class DoAfterDelayThread(threading.Thread):
+    '''A class to send signal.SIGUSR1 to cause EINTR in reading thread.'''
+    def __init__(self, initial_wait, func, *args, **kw):
+        self.initial_wait = initial_wait
+        self.func = func
+        super(DoAfterDelayThread, self).__init__(*args, **kw)
+
+    def run(self):
+        time.sleep(self.initial_wait)
+        self.func()
+
+class IOLoopEintrTestCaseSelect(IOLoopBaseTest):
+    ''' Tests if EINTR is properly caught and polling gets resumed. '''
+    SELECT_POLLER = 'select'
+    MSG_CONTENT = b'hello'
+
+    @staticmethod
+    def signal_handler(signum, interrupted_stack):
+        '''A signal handler that gets called in response to
+           os.kill(signal.SIGUSR1).'''
+        pass
+
+    def _eintr_read_handler(self, fileno, events, write_only):
+        '''Read from within poll loop that gets receives eintr error.'''
+        self.assertFalse(write_only)
+        self.assertEqual(events, READ)
+        sock = socket.fromfd(fileno, socket.AF_INET, socket.SOCK_STREAM)
+        mesg = sock.recv(256)
+        self.assertEqual(mesg, self.MSG_CONTENT)
+        self.poller.stop()
+        self._eintr_read_handler_is_called = True
+
+    def _eintr_test_fail(self):
+        '''This function gets called when eintr-test failed to get
+           _eintr_read_handler called.'''
+        self.poller.stop()
+        raise AssertionError('Eintr-test timed out')
+
+    #Disable warning about internal function access.
+    #pylint: disable=W0212
+    @PATCH_('pika.adapters.select_connection._is_resumable')
+    def test_eintr(self, is_resumable_mock, is_resumable_raw=
+                   pika.adapters.select_connection._is_resumable):
+        '''Test that poll() is properly restarted after receiving EINTR error.
+           Class of an exception raised to signal the error differs in one
+           implementation of polling mechanism and another.'''
+        if not pika.compat.HAVE_SIGNAL:
+            return
+        is_resumable_mock.side_effect = is_resumable_raw
+        self.poller = self.ioloop._get_poller()
+        sockpair = self.poller.get_interrupt_pair()
+        self._eintr_read_handler_is_called = False
+        self.poller.add_handler(sockpair[0].fileno(), self._eintr_read_handler,
+                                READ)
+        self.poller.add_timeout(self.TIMEOUT, self._eintr_test_fail)
+        assert getattr(self, 'original_signal_handler', None) is None
+        original_signal_handler = \
+            signal.signal(signal.SIGUSR1, self.signal_handler)
```

Comment:
It's more obvious and reliable to schedule cleanup right after registering the signal instead of having the cleanup code somewhere else:

`self.addCleanup(signal.signal, signal.SIGUSR1, original_signal_handler)`